### PR TITLE
Implement generic interposition effect.

### LINKED ISSUE #133: Define a zero-boilerplate generic interposition/eavesdropping carrier
Similar to #132’s proposed zero-boilerplate carrier for effects, we can define a zero-boilerplate interposition carrier for effects, allowing you to intercept and possibly respond to requests for other effects.

I’ve been adamantly opposed to adding other definitions of interposition which we’ve made, due to some of the egregious hacks they required, but this stab at it seems to work nicely while avoiding any such hackery:

```haskell
runInterpose :: (forall x . eff m (m x) -> m x) -> InterposeC eff m a -> m a
runInterpose handler = runReader (Handler handler) . runInterposeC

newtype InterposeC eff m a = InterposeC { runInterposeC :: ReaderC (Handler eff m) m a }
  deriving (Applicative, Functor, Monad)

instance MonadTrans (InterposeC eff) where
  lift = InterposeC . lift

newtype Handler eff m = Handler (forall x . eff m (m x) -> m x)

runHandler :: HFunctor eff => Handler eff m -> eff (ReaderC (Handler eff m) m) (ReaderC (Handler eff m) m a) -> m a
runHandler h@(Handler handler) = handler . handlePure (runReader h)

instance (HFunctor eff, Carrier sig m, Member eff sig) => Carrier sig (InterposeC eff m) where
  eff (op :: sig (InterposeC eff m) (InterposeC eff m a))
    | Just (op' :: eff (InterposeC eff m) (InterposeC eff m a)) <- prj op = do
      handler <- InterposeC ask
      lift (runHandler handler (handleCoercible op'))
    | otherwise = InterposeC (ReaderC (\ handler -> eff (handlePure (runReader handler . runInterposeC) op)))
```

Here’s a use of it to eavesdrop on requests for a `State` effect while still using the underlying `StateC` to respond:

```
λ runM (runState 5 (runInterpose @(State Integer) (\ op -> liftIO (putStrLn "state op") *> send op) get)) :: IO (Integer, Integer)
state op
( 5 , 5 )
```

And here’s a use of it to intercept and respond to requests for a `State` effect without involving the underlying `StateC` at all:

```
λ runM (runState 5 (runInterpose @(State Integer) (\case { Get k -> k 42 ; Put _ k -> k }) get)) :: IO (Integer, Integer)
( 5 , 42 )
```

This would supersede #6 (because we’d document this carrier itself).

### PR DESCRIPTION
**Pull request by patrickt**
Fixes #133.

### CODE REVIEW

File: src/Control/Effect/Interpose.hs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+-- and runs that handler whenever an effect of that type is encountered. Within a
+-- handler you can use all the capabilities of the underlying monad stack, including
+-- the intercepted effect, and you can pass the effect on to the original handler
+-- using 'send'.
+--
+-- The following shows the use of 'runInterpose' to eavesdrop on 'State' effects:
+-- @
+--  runM . runState @Int 5 . runInterpose @(State Int) (\op -> liftIO (print "State op") *> send op) $ go
+-- @
+runInterpose :: (forall x . eff m (m x) -> m x) -> InterposeC eff m a -> m a
```
Comment (robrix) [inline src/Control/Effect/Interpose.hs]: 🙏 doctest(s)?

File: src/Control/Effect/Interpose.hs [hunk 2]
```diff
@@ -0,0 +1,44 @@
+{-# LANGUAGE DeriveFunctor, ExplicitForAll, FlexibleContexts, FlexibleInstances, GeneralizedNewtypeDeriving, MultiParamTypeClasses, RankNTypes, ScopedTypeVariables, TypeOperators, UndecidableInstances #-}
+
+{- |
+This module provides an 'InterposeC' carrier capable of "eavesdropping" on requests
+made to other carriers. This is a useful capability for dynamism in deeply-nested
+effect stacks, but can lead to complicated control flow. Be careful.
+-}
+module Control.Effect.Interpose where
```
Comment (robrix) [inline src/Control/Effect/Interpose.hs]: Explicit export list, please 🙏

File: src/Control/Effect/Interpose.hs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  lift = InterposeC . lift
+
+newtype Handler eff m = Handler (forall x . eff m (m x) -> m x)
+
+runHandler :: HFunctor eff => Handler eff m -> eff (ReaderC (Handler eff m) m) (ReaderC (Handler eff m) m a) -> m a
+runHandler h@(Handler handler) = handler . handlePure (runReader h)
+
+instance (HFunctor eff, Carrier sig m, Member eff sig) => Carrier sig (InterposeC eff m) where
+  eff (op :: sig (InterposeC eff m) (InterposeC eff m a))
+    | Just (op' :: eff (InterposeC eff m) (InterposeC eff m a)) <- prj op = do
```
Comment (robrix) [inline src/Control/Effect/Interpose.hs]: At long last, _something_ uses `prj`!

File: src/Control/Effect/Interpose.hs (hunk 1 above)
Comment (patrickt) [reply src/Control/Effect/Interpose.hs]: Added.

File: src/Control/Effect/Interpose.hs [hunk 4]
```diff
@@ -18,10 +21,8 @@ import Control.Monad.Trans.Class
 -- the intercepted effect, and you can pass the effect on to the original handler
 -- using 'send'.
 --
--- The following shows the use of 'runInterpose' to eavesdrop on 'State' effects:
--- @
---  runM . runState @Int 5 . runInterpose @(State Int) (\op -> liftIO (print "State op") *> send op) $ go
--- @
+--   prop> run . evalState @Int a . runInterpose @(State Int) (\op -> modify @Int (+b) *> send op) $ modify @Int (+b) == a + b + b
```
Comment (robrix) [inline src/Control/Effect/Interpose.hs]: I’m not sure this is actually getting run, given the relative precedences of `==` and `$`. `doctest` doesn’t like telling you about errors, unfortunately 😞

**Pull request merged**